File: scraper/fetch_atom.js
import fetch from 'node-fetch';
import { Parser } from 'xml2js';
import { promises as fs } from 'fs';
import fsBase from 'fs';
import path from 'path';

const ATOM_FEED_URL = 'https://www.heise.de/rss/heise-atom.xml'; // Replace with your Atom feed URL

async function fetchAtomFeed(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch the Atom feed: ${response.statusText}`);
  }
  return response.text();
}

async function parseAtomFeed(xml) {
  const parser = new Parser();
  return new Promise((resolve, reject) => {
    parser.parseString(xml, (err, result) => {
      if (err) {
        reject(err);
      } else {
        resolve(result);
      }
    });
  });
}

async function getSavedIds(filePath) {
  try {
    const data = await fs.readFile(filePath, 'utf-8');
    return new Set(data.split('\n').filter(Boolean));
  } catch (err) {
    if (err.code === 'ENOENT') {
      return new Set(); // File doesn't exist, return empty set
    }
    throw err;
  }
}

async function saveIdAndUrl(id, url, title) {
  await fs.appendFile('fetch_atom.txt', `${id}\n`);
  await fs.appendFile('url.txt', `${id} ${url} ${title}\n`);
}

async function processAtomFeed() {
  try {
    const xml = await fetchAtomFeed(ATOM_FEED_URL);
    const parsedFeed = await parseAtomFeed(xml);

    const entries = parsedFeed.feed.entry || [];
    const savedIds = await getSavedIds(path.resolve('fetch_atom.txt'));

    let count = 0;

    for (const entry of entries) {
      const id = entry.id[0];
      const link = entry.link[0].$.href;
      const title = entry.title[0];

      if (!savedIds.has(id)) {
        await saveIdAndUrl(id, link, title['_']);
        count++;
      }
    }

    console.log('Done processing the Atom feed. Added %d new entries.', count);
  } catch (err) {
    console.error(`Error: ${err.message}`);
  }
}

function renameFileIfExists() {
  if (fsBase.existsSync('url.txt')) {
      let number = 1;
      let newFilePath;
      do {
          newFilePath = `url.txt.${number}`;
          number++;
      } while (fsBase.existsSync(newFilePath));
      fsBase.renameSync('url.txt', newFilePath);
      console.log(`Renamed 'url.txt' to '${newFilePath}'`);
  }
}

// Function to read file content and parse it into an array of IDs
function getIdsFromFile(filePath, columnIndex) {
  const fileContent = fsBase.readFileSync(filePath, 'utf-8');
  const lines = fileContent.split('\n');
  const ids = lines.map(line => {
      const columns = line.split(' ');
      return columns[columnIndex]?.trim();
  }).filter(Boolean); // Remove empty or undefined values
  return ids;
}

// Function to check if all IDs from one list exist in another
function checkIdsExist(sourceIds, targetIds) {
  return sourceIds.every(id => targetIds.includes(id));
}

if (fsBase.existsSync('url.txt')) {
  // Paths to the files
  const urlFilePath = 'url.txt';
  const pushToDBFilePath = 'pushToDB.txt';

  // Extract IDs from the files
  const urlIds = getIdsFromFile(urlFilePath, 0); // Extract from column 0 of url.txt
  const pushToDBIds = getIdsFromFile(pushToDBFilePath, 0); // Extract all lines from pushToDB.txt

  // Check if all IDs in url.txt are present in pushToDB.txt
  const allIdsExist = checkIdsExist(urlIds, pushToDBIds);

  if (allIdsExist) {
    renameFileIfExists();
    processAtomFeed();  
  } else {
    console.log('Some IDs from url.txt are missing in pushToDB.txt, thus not processed yet. Skipping to fetch more URLs.');
  }
} else {
  processAtomFeed();
}

File: scraper/fetch_atom.py


import requests
import xmltodict
import config
from loguru import logger

def fetch_atom_feed(url):
    response = requests.get(url)
    if not response.ok:
        raise Exception(f"Failed to fetch the Atom feed: {response.reason}")
    return response.text

def parse_atom_feed(xml):
    try:
        return xmltodict.parse(xml)
    except Exception as e:
        raise Exception(f"Error parsing XML: {e}")

def normalize_entries(parsed_feed):
    """
    Process Atom or RSS feeds into a unified structure.
    Returns a list of reference IDs and a mapping of ref_id -> entry details.
    """
    ref_ids = []
    entries_dict = {}

    def add_entry(_entry, _entry_id):
        if not _entry_id:
            return
        ref_ids.append(_entry_id)
        # Process link field
        link_obj = _entry.get('link')
        if isinstance(link_obj, list):
            link = link_obj[0].get('@href')
        elif isinstance(link_obj, dict):
            link = link_obj.get('@href')
        else:
            link = link_obj

        # Process title field
        title = _entry.get('title', '')
        if isinstance(title, dict):
            title = title.get('#text', '')
        entries_dict[_entry_id] = {'link': link, 'title': title}

    if 'feed' in parsed_feed and 'entry' in parsed_feed['feed']:
        entries = parsed_feed['feed']['entry']
        if not isinstance(entries, list):
            entries = [entries]
        for entry in entries:
            add_entry(entry, entry.get('id'))
    elif 'rss' in parsed_feed and 'channel' in parsed_feed['rss']:
        channels = parsed_feed['rss']['channel']
        if not isinstance(channels, list):
            channels = [channels]
        for channel in channels:
            items = channel.get('item', [])
            if not isinstance(items, list):
                items = [items]
            for item in items:
                guid = item.get('guid')
                entry_id = guid.get('#text') if isinstance(guid, dict) else guid
                add_entry(item, entry_id)
    return ref_ids, entries_dict


def perform_request(method, url, payload):
    response = requests.request(method, url, auth=(config.USERNAME, config.PASSWORD), json=payload)
    if not response.ok:
        raise Exception(f"Request failed: {response.status_code} {response.reason}")
    return response.json()

def filter_existing_entries(ref_ids):
    api_url = f"{config.URL}/api/v1/feed-item-to-process/filter"
    return perform_request("post", api_url, {'refIds': ref_ids})

def post_feed_item(ref_id, entry, feed_id):
    api_url = f"{config.URL}/api/v1/feed-item-to-process"
    payload = {
        'feedId': feed_id,
        'refId': ref_id,
        'url': entry['link'],
        'title': entry['title']
    }
    perform_request("post", api_url, payload)

def process_atom_feed(atom_feed_url, feed_id):
    xml = fetch_atom_feed(atom_feed_url)
    parsed_feed = parse_atom_feed(xml)
    ref_ids, entries = normalize_entries(parsed_feed)
    new_entries = filter_existing_entries(ref_ids)
    for ref_id in new_entries:
        if ref_id in entries:
            post_feed_item(ref_id, entries[ref_id], feed_id)
    if new_entries:
        logger.info(f"Done processing the Atom feed [{atom_feed_url}]. Added {len(new_entries)} new entries.")
